src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.cs

```diff
@@ -57,6 +61,10 @@ public void WriteDouble(double value)
         }
         private void WriteSingleCore(float value)
         {
+            if (float.IsNaN(value))
+            {
+                value = CanonicalNaNSingle;
+            }
```

Comment:
This isn't guaranteed to work across all possible runtimes/compilers/future versions. The IEEE 754 spec allows for a `NaN` to be normalized to any representation (and while we aren't likely to do that ourselves, it is very much legal to do so).

Rather, instead, you should have something like:
```csharp
if (float.IsNaN(value))
{
    WriteInt32Core(PositiveQNaNBits);
    return;
}

...
```

Reply:
WriteInt32Core doesn't exist but I agree, it shouldn't go through an FPU.